internal/cloud/generic_hostname.go

```diff
@@ -0,0 +1,66 @@
+package cloud
+
+import (
+	"log"
+	"net/url"
+
+	svchost "github.com/hashicorp/terraform-svchost"
+	"github.com/hashicorp/terraform-svchost/auth"
+	svcauth "github.com/hashicorp/terraform-svchost/auth"
+)
+
+var genericHostname = "localterraform.com"
+
+// configureGenericHost aliases the cloud backend hostname configuration
+// as a generic "localterraform.com" hostname. This was originally added as a
+// Terraform Enterprise feature and is useful for re-using whatever the
+// Cloud/Enterprise backend host is in nested module sources in order
+// to prevent code churn when re-using config between multiple
+// Terraform Enterprise environments.
+func (b *Cloud) configureGenericHostname() {
+	genericHost, _ := svchost.ForComparison(genericHostname)
+	aliasedCloudHost, _ := svchost.ForComparison(b.hostname)
+
+	// Alias the credentials from cloud host to generic host
+	creds, err := b.services.CredentialsForHost(aliasedCloudHost)
+	if err != nil {
+		log.Printf("[DEBUG] could not fetch creds: %s", err)
+		return
+	}
+
+	if creds != nil {
+		source := auth.StaticCredentialsSource(map[svchost.Hostname]map[string]interface{}{
+			genericHost: {
+				"token": creds.Token(),
+			},
+		})
+
+		// Replace current credentials source with a composite of both static
+		// credentials and current source
+		b.services.SetCredentialsSource(
+			svcauth.Credentials([]svcauth.CredentialsSource{
+				source,
+				b.services.CredentialsSource(),
+			}),
+		)
+	}
+
+	// Alias the cloud host as the generic host for these services
+	tfeURL := url.URL{
+		Scheme: "https",
+		Host:   aliasedCloudHost.String(),
+		Path:   "/api/v2/",
+	}
+
+	modulesURL := url.URL{
+		Scheme: "https",
+		Host:   aliasedCloudHost.String(),
+		Path:   "/api/registry/v1/modules/",
+	}
+
+	log.Printf("[TRACE] configured generic host %q as an alias for %q", genericHost, aliasedCloudHost)
+	b.services.ForceHostServices(genericHost, map[string]interface{}{
+		"modules.v1": modulesURL.String(),
+		"tfe.v2":     tfeURL.String(),
```

Comment:
This is not overriding the "providers.v1" service, used when fetching providers. I believe this is also a different `*disco.Disco` instance from the one used by the `getproviders` package, although I haven't dug into that any further.

Have you been able to test this locally, and verify that this change works for providers in the private registry?

Reply:
Hmm this is a good point, and also raises another point: this is locking in place a fixed set of services currently offered by Terraform Cloud instances, which means it won't be able to see new services added in future and we won't be able to change the URLs of any existing services for Terraform Cloud.

Although I concede that this will require deeper changes to the "disco" abstraction itself, I wonder if we should teach that layer about the idea of "host aliases" so we can tell it "whenever someone asks about `localterraform.com`, pretend that they asked for `app.terraform.io` and then still do exactly the same network lookup to fetch the services for that real host.

I think this also echoes how we're treating a credentials lookup for `localterraform.com` the same as one for whatever real hostname that's an alias for.

I've not looked closely to see how feasible that would be; perhaps there's another variation that would be similar.

Reply:
@apparentlymart I had the same thought yesterday but disregarded it in my haste. I think your idea of making this a feature of svchost at the host level simplifies everything about change.

Reply:
@apparentlymart [this](https://github.com/hashicorp/terraform-svchost/pull/20) seems to work nicely and doesn't suffer from anticipating the exact services within discovery.

Reply:
That change looks like it'll address the problem with the service name, but not the other issue: this is not the same `*disco.Disco` instance used to fetch providers.

Here's a starting point for pursuing that:

https://github.com/hashicorp/terraform/blob/7d2afaa2afbc3f6f2e63f428cdd53eccf31b6d16/main.go#L141-L162

Unless I've missed something, we'd need to update that one too. Have you been able to test this against a TFE instance to check that it correctly installs providers from `localterraform.com`?

Reply:
This appears to be the same services discovery object. I believe this pointer is passed into initCommands and ultimately the Meta struct for each command.

```
Initializing provider plugins...
------> - Finding localterraform.com/hashicorp/nested versions matching "1.2.0"...
------> 2023-01-26T14:52:06.125-0700 [DEBUG] Discover found alias localterraform.com for tfcdev-5839ba17.ngrok.io
2023-01-26T14:52:06.125-0700 [DEBUG] Service discovery for tfcdev-5839ba17.ngrok.io at https://tfcdev-5839ba17.ngrok.io/.well-known/terraform.json
2023-01-26T14:52:06.125-0700 [TRACE] HTTP client GET request to https://tfcdev-5839ba17.ngrok.io/.well-known/terraform.json
2023-01-26T14:52:07.588-0700 [DEBUG] CredentialsForHost found alias localterraform.com for tfcdev-5839ba17.ngrok.io
------> 2023-01-26T14:52:07.588-0700 [DEBUG] GET https://tfcdev-5839ba17.ngrok.io/api/registry/v1/providers/hashicorp/nested/versions
```

Reply:
Oh, great! My mistake.